python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -0,0 +1,189 @@
+    config
+    return
+
+
+@PublicAPI(stability="alpha")
+class HuggingFaceTrainer(TorchTrainer):
+    """A Trainer for data parallel HuggingFace Transformers on PyTorch training.
+
+    This Trainer runs the ``transformers.Trainer.train()`` method on multiple
+    Ray Actors. The training is carried out in a distributed fashion through PyTorch
+    DDP. These actors already have the necessary torch process group already
+    configured for distributed PyTorch training.
+
+    The training function ran on every Actor will first run the
+    specified ``trainer_init_per_worker`` function to obtain an instantiated
+    ``transformers.Trainer`` object. The ``trainer_init_per_worker`` function
+    will have access to preprocessed train and evaluation datsets.
+
+    If the ``datasets`` dict contains a training dataset (denoted by
+    the "train" key), then it will be split into multiple dataset
+    shards, with each Actor training on a single shard.
+    All the other datasets will not be split.
+
+    Please note that if you use a custom ``transformers.Trainer`` subclass,
+    the ``get_train_dataloader`` method will be overriden to disable distributed
+    sampling, as the dataset will already be sharded.
+
+    Hugging Face loggers will be automatically disabled, and the ``local_rank``
+    argument in ``TrainingArguments`` will be automatically set.
+
+    Example:
+        .. code-block:: python
+            # Based on
+            # huggingface/notebooks/examples/language_modeling_from_scratch.ipynb
+
+            # Hugging Face imports
+            from datasets import load_dataset
+            from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
+            from transformers import Trainer, TrainingArguments
+
+            import ray
+            from ray.ml.train.integrations.huggingface import HuggingFaceTrainer
+
+            model_checkpoint = "gpt2"
+            tokenizer_checkpoint = "sgugger/gpt2-like-tokenizer"
+            block_size = 128
+
+            datasets = load_dataset("wikitext", "wikitext-2-raw-v1")
+            tokenizer = AutoTokenizer.from_pretrained(tokenizer_checkpoint)
+
+            def tokenize_function(examples):
+                return tokenizer(examples["text"])
+
+            tokenized_datasets = datasets.map(
+                tokenize_function, batched=True, num_proc=1, remove_columns=["text"]
+            )
+
+            def group_texts(examples):
+                # Concatenate all texts.
+                concatenated_examples = {
+                    k: sum(examples[k], []) for k in examples.keys()
+                }
+                total_length = len(concatenated_examples[list(examples.keys())[0]])
+                # We drop the small remainder, we could add padding if the model
+                # supported it.
+                # instead of this drop, you can customize this part to your needs.
+                total_length = (total_length // block_size) * block_size
```

Comment:
you don't really have to do this explicitly, since range(0, total_length, block_size) effectively drops the reminder for you?

also usability nits:
can we give folks some ideas about how to customize this part without dropping the last block? e.g., compute a right block_size automatically so minimum amount of samples get dropped? would the last block being slightly smaller still work for torch ddp?

Reply:
I see. ok.